src/lib/transition_router/initial_validator.ml

```diff
@@ -46,24 +47,73 @@ module Make (Inputs : Transition_frontier.Inputs_intf) = struct
              ( "off by $slot_diff slots"
              , [("slot_diff", `String (Int64.to_string slot_diff))] ))
 
+  module Duplicate_proposal_detector = struct
+    module T = struct
+      (* int is a slot *)
+      type t = Public_key.Compressed.Stable.V1.t * int
+      [@@deriving bin_io, sexp, hash, compare]
+    end
+
+    module Key = Hashable.Make_binable (T)
+
+    type t = {table: State_hash.t Key.Table.t; mutable latest_epoch: int}
```

Comment:
For each proposer, can the duplicated proposal be stored in a ring buffer rather than a Hashtbl?

Reply:
I started down that road, a little ways. I have a couple of concerns with that. 

The advantage of a ring buffer would be to limit the storage required. While there is a notion of "network delay", which might restrict the range of slots in play, so that the ring buffer is correspondingly limited in size, I'm worried that network latency might mean seeing transitions with slots outside of that range, such that there'd be no corresponding location in the ring buffer.

The second concern is complexity of implementation. I started implementing that, adjusting the buffer bounds as you see new transitions. There might, in fact, be a lot of wasted space if there are lots of proposers, but each only proposes a small number of slots, so that most entries in the buffers are unused.